// jest-metadata/dist/metadata/symbols.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pushAfterAll = exports.pushBeforeAll = exports.pushExecution = exports.start = exports.rootDescribeBlock = exports.id = exports.finish = exports.data = exports.reportedTestEntries = exports.lastTestEntry = exports.currentMetadata = exports.context = exports.addTestEntry = exports.addHookDefinition = exports.addDescribeBlock = void 0;
exports.addDescribeBlock = Symbol('ADD_DESCRIBE_BLOCK');
exports.addHookDefinition = Symbol('ADD_HOOK_DEFINITION');
exports.addTestEntry = Symbol('ADD_TEST_ENTRY');
exports.context = Symbol('CONTEXT');
exports.currentMetadata = Symbol('CURRENT_METADATA');
exports.lastTestEntry = Symbol('LAST_TEST_ENTRY');
exports.reportedTestEntries = Symbol('REPORTED_TEST_ENTRIES');
exports.data = Symbol('DATA');
exports.finish = Symbol('FINISH');
exports.id = Symbol('ID');
exports.rootDescribeBlock = Symbol('ROOT_DESCRIBE_BLOCK');
exports.start = Symbol('START');
exports.pushExecution = Symbol('PUSH_EXECUTION');
exports.pushBeforeAll = Symbol('PUSH_BEFORE_ALL');
exports.pushAfterAll = Symbol('PUSH_AFTER_ALL');
