// jest-metadata/dist/utils/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nologger = exports.diagnostics = exports.logger = exports.optimizeTracing = void 0;
const bunyamin_1 = require("bunyamin");
function isTraceEnabled() {
    return (0, bunyamin_1.isDebug)('jest-metadata');
}
bunyamin_1.threadGroups
    .add({ id: 'jest-metadata', displayName: 'jest-metadata' })
    .add({
    id: 'jest-metadata-emitter-core',
    displayName: 'jest-metadata (Core Emitter)',
})
    .add({
    id: 'jest-metadata-emitter-events',
    displayName: 'jest-metadata (Events Emitter)',
})
    .add({
    id: 'jest-metadata-emitter-globalMetadataRegistry',
    displayName: 'jest-metadata (Metadata Registry)',
})
    .add({ id: 'jest-metadata-emitter-set', displayName: 'jest-metadata (Set Emitter)' })
    .add({ id: 'jest-metadata-environment', displayName: 'jest-metadata (Environment)' })
    .add({ id: 'jest-metadata-ipc', displayName: 'jest-metadata (IPC)' })
    .add({ id: 'jest-metadata-reporter', displayName: 'jest-metadata (Reporter)' });
const EMPTY = {};
const NOOP = () => EMPTY;
exports.optimizeTracing = isTraceEnabled() ? (f) => f : () => NOOP;
exports.logger = bunyamin_1.bunyamin.child({ cat: 'jest-metadata' });
exports.diagnostics = isTraceEnabled() ? exports.logger : bunyamin_1.nobunyamin;
var bunyamin_2 = require("bunyamin");
Object.defineProperty(exports, "nologger", { enumerable: true, get: function () { return bunyamin_2.nobunyamin; } });
