// jest-allure2-reporter/dist/runtime/modules/CoreModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoreModule = void 0;
class CoreModule {
    context;
    constructor(context) {
        this.context = context;
    }
    static create(context) {
        return new CoreModule({
            get metadata() {
                return context.getCurrentMetadata();
            },
        });
    }
    // region Universal (test, hook, step) metadata
    displayName(value) {
        this.context.metadata.set('displayName', value);
    }
    parameter(parameter) {
        this.context.metadata.push('parameters', [parameter]);
    }
    status(status) {
        this.context.metadata.set('status', status);
    }
    statusDetails(statusDetails) {
        this.context.metadata.set('statusDetails', statusDetails);
    }
    // endregion
    // region Test-only metadata
    description(value) {
        this.context.metadata.$bind(null).push('description', [value]);
    }
    descriptionHtml(value) {
        this.context.metadata.$bind(null).push('descriptionHtml', [value]);
    }
    fullName(value) {
        this.context.metadata.$bind(null).set('fullName', value);
    }
    historyId(value) {
        this.context.metadata.$bind(null).set('historyId', value);
    }
    label(name, value) {
        this.context.metadata.$bind(null).push('labels', [{ name, value }]);
    }
    link(link) {
        this.context.metadata.$bind(null).push('links', [link]);
    }
}
exports.CoreModule = CoreModule;
