// jest-metadata/dist/debug/plant/MetadataPlantObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataPlantObject = void 0;
const PlantObject_1 = require("./PlantObject");
const COLORS = {
    GlobalMetadata: '#dff',
    TestFileMetadata: '#def',
    DescribeBlockMetadata: '#ded',
    HookDefinitionMetadata: '#fdd',
    HookInvocationMetadata: '#fcb',
    TestEntryMetadata: '#ffd',
    TestInvocationMetadata: '#ffb',
    TestFnInvocationMetadata: '#fea',
};
class MetadataPlantObject extends PlantObject_1.PlantObject {
    constructor(metadata) {
        const name = Object.getPrototypeOf(metadata).constructor.name;
        super({
            id: metadata.id,
            name,
            color: COLORS[name],
            properties: getProperties(metadata),
        });
    }
}
exports.MetadataPlantObject = MetadataPlantObject;
function getProperties(metadata) {
    const [fileId, childId] = metadata.id.split(':');
    const result = {
        id: childId || fileId,
        data: metadata.get(),
    };
    if (!result.id) {
        delete result.id;
    }
    if (!result.data || Object.keys(result.data).length === 0) {
        delete result.data;
    }
    return result;
}
