// jest-allure2-reporter/dist/options/common/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.merger = exports.mapper = exports.composite = exports.last = exports.constant = exports.compose3 = exports.compose2 = exports.appender = void 0;
var appenderExtractor_1 = require("./appenderExtractor");
Object.defineProperty(exports, "appender", { enumerable: true, get: function () { return appenderExtractor_1.appenderExtractor; } });
var composeExtractors2_1 = require("./composeExtractors2");
Object.defineProperty(exports, "compose2", { enumerable: true, get: function () { return composeExtractors2_1.composeExtractors2; } });
var composeExtractors3_1 = require("./composeExtractors3");
Object.defineProperty(exports, "compose3", { enumerable: true, get: function () { return composeExtractors3_1.composeExtractors3; } });
var constantExtractor_1 = require("./constantExtractor");
Object.defineProperty(exports, "constant", { enumerable: true, get: function () { return constantExtractor_1.constantExtractor; } });
var last_1 = require("./last");
Object.defineProperty(exports, "last", { enumerable: true, get: function () { return last_1.last; } });
var compositeExtractor_1 = require("./compositeExtractor");
Object.defineProperty(exports, "composite", { enumerable: true, get: function () { return compositeExtractor_1.compositeExtractor; } });
var mapper_1 = require("./mapper");
Object.defineProperty(exports, "mapper", { enumerable: true, get: function () { return mapper_1.mapper; } });
var mergerExtractor_1 = require("./mergerExtractor");
Object.defineProperty(exports, "merger", { enumerable: true, get: function () { return mergerExtractor_1.mergerExtractor; } });
