// jest-metadata/dist/realms/ParentProcessRealm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParentProcessRealm = void 0;
const ipc_1 = require("../ipc");
const jest_reporter_1 = require("../jest-reporter");
const utils_1 = require("../utils");
const BaseRealm_1 = require("./BaseRealm");
const detect_1 = require("./detect");
class ParentProcessRealm extends BaseRealm_1.BaseRealm {
    type = 'parent_process';
    ipc = new ipc_1.IPCServer({
        appspace: `jest-metadata@${(0, utils_1.getVersion)()}-`,
        serverId: `${process.pid}`,
        emitter: this.coreEmitter,
        globalMetadata: this.globalMetadata,
    });
    fallbackAPI = new jest_reporter_1.FallbackAPI(this.globalMetadata, this.coreEmitter);
    reporterServer = new jest_reporter_1.ReporterServer({
        associate: this.associate,
        fallbackAPI: this.fallbackAPI,
        ipc: this.ipc,
        rootDir: process.cwd(),
    });
    constructor() {
        super();
        (0, detect_1.registerServerId)(this.ipc.id);
        this.associate.globalMetadata(this.globalMetadata);
    }
}
exports.ParentProcessRealm = ParentProcessRealm;
