// jest-allure2-reporter/dist/options/custom/labels/simplifyLabelsMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.simplifyLabelsMap = simplifyLabelsMap;
const utils_1 = require("../../../utils");
const common_1 = require("../../common");
const keyedLabel_1 = require("./keyedLabel");
function simplifyLabelsMap(customizer) {
    return (0, utils_1.compactObject)((0, utils_1.mapValues)(customizer, (value, key) => {
        const keyedCustomizer = (0, keyedLabel_1.keyedLabel)(value);
        return keyedCustomizer
            ? (0, common_1.compose3)(inflateLabels(key), keyedCustomizer, deflateLabels)
            : undefined;
    }));
}
function deflateLabels(context) {
    return (0, utils_1.asMaybeArray)((0, utils_1.asArray)(context.value).map(getValue));
}
function getValue(label) {
    return label.value;
}
function inflateLabels(name) {
    function repair(value) {
        return value ? { value, name } : undefined;
    }
    function repairMaybeArray(value) {
        return (0, utils_1.compactArray)((0, utils_1.asArray)(value).map(repair));
    }
    return ({ value }) => (0, utils_1.thruMaybePromise)(value, repairMaybeArray);
}
