// jest-allure2-reporter/dist/utils/vendor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uniq = exports.snakeCase = exports.once = exports.memoize = exports.merge = exports.mapValues = exports.last = exports.isObject = exports.isError = exports.isEmpty = exports.groupBy = exports.get = exports.defaults = exports.constant = exports.compactArray = exports.attempt = void 0;
const tslib_1 = require("tslib");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
exports.attempt = lodash_1.default.attempt;
exports.compactArray = lodash_1.default.compact;
exports.constant = lodash_1.default.constant;
exports.defaults = lodash_1.default.defaults;
exports.get = lodash_1.default.get;
exports.groupBy = lodash_1.default.groupBy;
exports.isEmpty = lodash_1.default.isEmpty;
exports.isError = lodash_1.default.isError;
exports.isObject = lodash_1.default.isObject;
exports.last = lodash_1.default.last;
exports.mapValues = lodash_1.default.mapValues;
exports.merge = lodash_1.default.merge;
exports.memoize = lodash_1.default.memoize;
exports.once = lodash_1.default.once;
exports.snakeCase = lodash_1.default.snakeCase;
exports.uniq = lodash_1.default.uniq;
