// jest-allure2-reporter/dist/options/custom/testCase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testCase = testCase;
const tslib_1 = require("tslib");
const extractors = tslib_1.__importStar(require("../common"));
const custom = tslib_1.__importStar(require("../custom"));
const parameters_1 = require("./parameters");
const fallback = (context) => context.value;
function testCase(customizer) {
    if (customizer == null) {
        return;
    }
    return extractors.composite({
        uuid: extractors.constant(customizer.uuid) ?? fallback,
        ignored: extractors.constant(customizer.ignored) ?? fallback,
        historyId: extractors.constant(customizer.historyId) ?? fallback,
        displayName: extractors.constant(customizer.displayName) ?? fallback,
        fullName: extractors.constant(customizer.fullName) ?? fallback,
        start: extractors.constant(customizer.start) ?? fallback,
        stop: extractors.constant(customizer.stop) ?? fallback,
        description: extractors.constant(customizer.description) ?? fallback,
        descriptionHtml: extractors.constant(customizer.descriptionHtml) ?? fallback,
        stage: extractors.constant(customizer.stage) ?? fallback,
        status: extractors.constant(customizer.status) ?? fallback,
        statusDetails: extractors.merger(customizer.statusDetails) ?? fallback,
        attachments: extractors.appender(customizer.attachments) ?? fallback,
        parameters: (0, parameters_1.parameters)(customizer.parameters) ?? fallback,
        labels: custom.labels(customizer.labels) ?? fallback,
        links: custom.links(customizer.links) ?? fallback,
        steps: null,
    });
}
