// jest-allure2-reporter/dist/options/custom/testStep.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testStep = testStep;
const tslib_1 = require("tslib");
const extractors = tslib_1.__importStar(require("../common"));
const parameters_1 = require("./parameters");
const fallback = (context) => context.value;
function testStep(customizer) {
    if (customizer == null) {
        return;
    }
    return extractors.composite({
        hookType: fallback,
        steps: null,
        ignored: extractors.constant(customizer.ignored) ?? fallback,
        displayName: extractors.constant(customizer.displayName) ?? fallback,
        start: extractors.constant(customizer.start) ?? fallback,
        stop: extractors.constant(customizer.stop) ?? fallback,
        stage: extractors.constant(customizer.stage) ?? fallback,
        status: extractors.constant(customizer.status) ?? fallback,
        statusDetails: extractors.merger(customizer.statusDetails) ?? fallback,
        attachments: extractors.appender(customizer.attachments) ?? fallback,
        parameters: (0, parameters_1.parameters)(customizer.parameters) ?? fallback,
    });
}
